Realm/RLMObjectInfo.hpp

```diff
@@ -0,0 +1,102 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#import <Foundation/Foundation.h>
+#import <unordered_map>
+#import <vector>
+
+namespace realm {
+    class ObjectSchema;
+    class Schema;
+    class Table;
+    struct Property;
+}
+
+class RLMObservationInfo;
+@class RLMRealm, RLMSchema, RLMObjectSchema, RLMProperty;
+
+NS_ASSUME_NONNULL_BEGIN
+
+namespace std {
+// Add specializations so that NSString can be used as the key for hash containers
+template<> struct hash<NSString *> {
+    size_t operator()(__unsafe_unretained NSString *const str) const {
+        return [str hash];
+    }
+};
+template<> struct equal_to<NSString *> {
+    bool operator()(__unsafe_unretained NSString * lhs, __unsafe_unretained NSString *rhs) const {
+        return [lhs isEqualToString:rhs];
+    }
+};
+}
+
+// The per-RLMRealm object schema information which stores the cached table
+// reference, handles table column lookups, and tracks observed objects
+struct RLMObjectInfo {
```

Comment:
Any suggestions for better names? `RLMObjectSchemaInfo` is the only obvious one, but it feels kinda gross to me.